src/cargo/ops/cargo_rustc/mod.rs

```diff
@@ -266,7 +270,36 @@ fn rustc(cx: &mut Context, unit: &Unit) -> CargoResult<Work> {
         }
 
         state.running(&rustc);
-        try!(exec_engine.exec(rustc).chain_error(|| {
+        let process_builder = rustc.into_process_builder();
+        try!(if json_errors {
+            #[derive(RustcEncodable)]
+            struct Message<'a> {
```

Comment:
Perhaps this could be defined in a more central location? This'd in theory I guess actually be some sort of enum for the messages that get emitted, and all of them have a `reason`.